### DISCUSSION TITLE
有关汉字读写问题

### DISCUSSION THREAD

**Discussion by WT1W** [liusy58/NJU_AdvancedProgramming_SP22#52] [Q&A] [answered]
txt文件中有汉字的时候，将文件中的内容读到字符串中，再将字符串进行输出，会出现中文乱码的问题，这个是因为GBK编码造成的问题，这个情况除了修改txt的编码方式以外还有其他解决方法吗？如果用更改txt文件编码（比如修改为ANSI）进行解决，那么验收的时候可以现场手动更改txt的编码方式吗？另外这种方式如果要移植到linux上可行吗？

**Comment by syhien** [comment]
可以浅了解一下`std::wstring`

**Reply by WT1W** [reply]
那可以通过修改txt编码属性实现吗，验收的时候会有问题吗

**Reply by liusy58** [reply]
个人认为用utf-8编码可以解决很多问题，网上应该也有很多的参考实现。

**Comment by cqjjjzr** [comment]
建议是 UTF-8 化。在 Windows 上需要三方面：

- 代码文件 UTF-8 化：Visual Studio 需要修改 `.editorconfig` 文件，见 https://stackoverflow.com/a/65945041/7732043
- 代码编译 UTF-8 化：Linux 上不需要特殊处理，Windows （MSVC）上需要打开 `/utf-8` 编码开关
- 控制台切换 UTF-8：使用 `system("chcp 65001")` 或 `SetConsoleCP` 与 `SetConsoldOutputCP` 两个函数。

然后看看本 repo `utf8_console` 目录 下的东西

**Reply by cqjjjzr** [reply]
同时 Linux 下现在默认都是 UTF-8 了，如果没有特意改 `LOCALE`，则可以假定控制台都是 UTF-8 的

**Reply by laz7cat** [reply]
正解!昨天折腾了一下就是这么解决的

**Reply by JusenD** [reply]
文件输入输出解决了，但是在控制台输入的中文还是不能utf-8化怎么办

**Reply by Price1999a** [reply]
> 文件输入输出解决了，但是在控制台输入的中文还是不能utf-8化怎么办

假如你真遇到困难 考虑实现全英文版本？

**Comment by bliss-bug** [comment]
一般来说在windows上，如果你向文件里面传入了汉字，默认编码格式是ANSI

**Comment by RicoloveFeng** [comment, marked as answer]
[见#66这里](https://github.com/liusy58/NJU_AdvancedProgramming_SP22/discussions/66#discussioncomment-2244767)
